File: frontend/heroic/heroic.js
/*!
 * Isomer - The distributed application framework
 * ==============================================
 * Copyright (C) 2011-2019 Heiko 'riot' Weinen <riot@c-base.org> and others.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

'use strict';

class countablescomponent {

    constructor(objectproxy, user, $state, $scope, socket) {
        this.op = objectproxy;
        this.user = user;
        this.state = $state;
        this.scope = $scope;
        this.socket = socket;

        let self = this;

        self.achievements = [];
        self.badges = [];

        this.getAwards = function () {
            self.op.search('achievements', '', '*').then(function (msg) {
                let achievements = msg.data.list;
                console.log('[HERO] Achievements:', achievements);
                for (let item of achievements) {
                    self.achievements.push(item);
                }
            });
            self.op.search('badges', '', '*').then(function (msg) {
                let badges = msg.data.list;
                for (let item of badges) {
                    self.badges.push(item);
                }
            });

        };

        if (this.user.signedin === true) {
            console.log('User signed in. Getting data.');
            this.getAwards();
        } else {
            console.log('Not logged in apparently, heres this:', this.user);
        }

        $scope.$on('User.Login', this.getAwards);
    }
}

countablescomponent.$inject = ['objectproxy', 'user', '$state', '$scope', 'socket'];

export default countablescomponent;

File: isomer/heroic/heroic.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

# Isomer - The distributed application framework
# ==============================================
# Copyright (C) 2011-2019 Heiko 'riot' Weinen <riot@c-base.org> and others.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

__author__ = "Heiko 'riot' Weinen"
__license__ = "AGPLv3"

"""


Module: Heroic
==============

Listens to events and adds achievements accordingly


"""

from isomer.component import ConfigurableComponent, handler
from isomer.database import objectmodels
from isomer.logger import isolog, error, warn, critical, events
from isomer.events.system import authorized_event
from pprint import pprint


class Heroic(ConfigurableComponent):
    """
    Watches for authorized events and adds achievements to a user's profile.
    """
    channel = 'isomer-web'

    configprops = {
    }

    def __init__(self, *args):
        """
        Initialize the Heroic component.

        :param args:
        """

        super(Heroic, self).__init__("HEROIC", *args)

        self.log("Started")
        self.accounts = {}

        for account in objectmodels['user'].find():
            self.accounts[account.uuid] = account

        self.log('Cached', len(self.accounts), 'User accounts')

    # @handler()
    # def authorized_event_handler(self, event, *args, **kwargs):
    #    self.log(event, args, kwargs)
    # if isinstance(event, authorizedevent):
    #    self.log('Event:', event)
    #    account = self.accounts[event.useruuid]
    #    self.log('Account', account)
